File: app/src/main/java/com/shafa/ali/kavir_msg/activity/CategoryActivity.java
package com.shafa.ali.kavir_msg.activity;

import android.app.Fragment;
import android.app.FragmentTransaction;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.Typeface;
import android.support.annotation.NonNull;
import android.support.design.widget.FloatingActionButton;
import android.support.design.widget.NavigationView;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBarDrawerToggle;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.text.Spannable;
import android.text.SpannableString;
import android.util.Log;
import android.view.GestureDetector;
import android.view.Gravity;
import android.view.Menu;
import android.view.MenuItem;
import android.view.MotionEvent;
import android.view.SubMenu;
import android.view.View;
import android.widget.ImageButton;
import android.widget.TextView;
import android.widget.Toast;

import com.github.ybq.android.spinkit.SpinKitView;
import com.shafa.ali.kavir_msg.R;
import com.shafa.ali.kavir_msg.adapters.CategoryAdapter;
import com.shafa.ali.kavir_msg.fragments.AboutFragment;
import com.shafa.ali.kavir_msg.fragments.AccessFragment;
import com.shafa.ali.kavir_msg.fragments.ActiveFragment;
import com.shafa.ali.kavir_msg.fragments.CategoryFragment;
import com.shafa.ali.kavir_msg.fragments.ReadyReadFragment;
import com.shafa.ali.kavir_msg.fragments.RegisterFragment;
import com.shafa.ali.kavir_msg.interfaces.ClickListener;
import com.shafa.ali.kavir_msg.models.CategoryModel;
import com.shafa.ali.kavir_msg.server.GetDataCategory;
import com.shafa.ali.kavir_msg.utility.CustomTypeFaceSpan;
import com.shafa.ali.kavir_msg.utility.FormatHelper;
import com.shafa.ali.kavir_msg.utility.RetrofitClientInstance;
import com.shafa.ali.kavir_msg.utility.SaveItem;
import com.shafa.ali.kavir_msg.utility.Utility;
import com.valdesekamdem.library.mdtoast.MDToast;

import java.util.List;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class CategoryActivity extends AppCompatActivity implements NavigationView.OnNavigationItemSelectedListener {
    private Toolbar mTopToolbar;
    private FragmentTransaction transaction;
    private DrawerLayout drawer;
    private TextView toolbarTitle,displayNameTv;
    private NavigationView navigationView;
    private ImageButton searchBtn;

    public static Context context;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_category);
        context = this;
        Log.e("mac    ",Utility.getMacAddr());
        mTopToolbar = (Toolbar) findViewById(R.id.my_toolbar);
        toolbarTitle =(TextView) findViewById(R.id.category_title);
        searchBtn =(ImageButton) findViewById(R.id.search_btn);

        searchBtn.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                startActivity(new Intent(CategoryActivity.this,SearchActivity.class));
            }
        });
        setSupportActionBar(mTopToolbar);

        //***************************
        drawer = (DrawerLayout) findViewById(R.id.drawer_layout);
        ActionBarDrawerToggle toggle = new ActionBarDrawerToggle(
                this, drawer, mTopToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close);

        navigationView = (NavigationView) findViewById(R.id.nav_view);
        View headerView = navigationView.getHeaderView(0);
        displayNameTv =(TextView)headerView.findViewById(R.id.display_name_header);
        String displayName  =SaveItem.getItem(this,SaveItem.USER_NAME,"");
        if (displayName.equals("")){
            displayName = getString(R.string.guest);
        }
        displayNameTv.setText(FormatHelper.toPersianNumber(displayName));
        navigationView.setNavigationItemSelectedListener(this);
        setNameLoginItem();
        toggle.setDrawerIndicatorEnabled(false);
        drawer.addDrawerListener(toggle);

        toggle.syncState();
        //////***************************************
        findViewById(R.id.drawer_button).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                // open right drawer

                if (drawer.isDrawerOpen(GravityCompat.END)) {
                    drawer.closeDrawer(GravityCompat.END);
                } else
                    drawer.openDrawer(GravityCompat.END);
            }
        });

        setFontNavig();


    }

    @Override
    protected void onResume() {
        try{
            CategoryFragment fragment = (CategoryFragment) getFragmentManager().findFragmentById(R.id.fragment);
            fragment.getDataFromServer();
        }catch (Exception e){
            Log.e("onResume:",e.getMessage());
        }

        super.onResume();
    }

    private void setFontNavig() {
        Menu m = navigationView.getMenu();
        for (int i=0;i<m.size();i++) {
            MenuItem mi = m.getItem(i);

            //for aapplying a font to subMenu ...
            SubMenu subMenu = mi.getSubMenu();
            if (subMenu!=null && subMenu.size() >0 ) {
                for (int j=0; j <subMenu.size();j++) {
                    MenuItem subMenuItem = subMenu.getItem(j);
                    applyFontToMenuItem(subMenuItem);
                }
            }
            applyFontToMenuItem(mi);
        }
    }

    private void loadFragment(Fragment fragmentNew) {
        transaction = getFragmentManager().beginTransaction();
        Fragment fragment = getFragmentManager().findFragmentById(R.id.fragment);
        if (fragment!=null) {
            getFragmentManager().beginTransaction().remove(fragment).commit();
        }
        // load fragment
        transaction.replace(R.id.fragment, fragmentNew);
        transaction.addToBackStack(null);
        transaction.commit();
    }



    @Override
    public void onBackPressed() {

        if (drawer.isDrawerOpen(GravityCompat.END)) {
            drawer.closeDrawer(GravityCompat.END);
        } else {
            super.onBackPressed();
        }
    }



    @Override
    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
        // Handle navigation view item clicks here.
        int id = menuItem.getItemId();

        if (id == R.id.login) {
            loginProcess();
        } else if (id == R.id.home_page) {
            loadFragment(CategoryFragment.newInstance());
            toolbarTitle.setText(getString(R.string.home));
        } else if (id == R.id.ready_read_page) {
            loadFragment(ReadyReadFragment.newInstance());
            toolbarTitle.setText(getString(R.string.redy_read));
        }else if (id == R.id.about_we){
            loadFragment(AboutFragment.newInstance());
            toolbarTitle.setText(getString(R.string.about_we));
        }else if (id == R.id.register){
            loadFragment(RegisterFragment.newInstance());
            toolbarTitle.setText(getString(R.string.register));
        }else if (id==R.id.active){
            loadFragment(ActiveFragment.newInstance());
            toolbarTitle.setText(getString(R.string.active));
        }else if (id==R.id.access){
            loadFragment(AccessFragment.newInstance());
            toolbarTitle.setText(getString(R.string.access_list));
        }

        drawer.closeDrawer(Gravity.END);
        return true;
    }

    private void loginProcess() {
        Menu menu = navigationView.getMenu();
        MenuItem loginItem = menu.findItem(R.id.login);
        if (loginItem.getTitle().toString().equalsIgnoreCase(getString(R.string.login))){
            startActivity(new Intent(CategoryActivity.this,LoginActivity.class));
        }else{
            logout();
        }
    }

    private void logout() {
        clearSeesion();
        finish();
    }

    private void clearSeesion(){
        SaveItem.setItem(this,SaveItem.USER_FIRST_NAME,"");
        SaveItem.setItem(this,SaveItem.USER_LAST_NAME,"");
        SaveItem.setItem(this,SaveItem.USER_EMAIL,"");
        SaveItem.setItem(this,SaveItem.USER_NAME,"");
        SaveItem.setItem(this,SaveItem.USER_MOBILE,"");
        SaveItem.setItem(this,SaveItem.USER_ID,"");
        SaveItem.setItem(this,SaveItem.USER_COOKIE,"");
    }

    private void setNameLoginItem(){
        Menu menu = navigationView.getMenu();
        MenuItem loginItem = menu.findItem(R.id.login);
        MenuItem activeItem= menu.findItem(R.id.active);
        MenuItem registerItem = menu.findItem(R.id.register);
        MenuItem accessItem = menu.findItem(R.id.access);
        if (SaveItem.getItem(this,SaveItem.USER_COOKIE,"").equals("")){
            loginItem.setTitle(getString(R.string.login));
            activeItem.setVisible(false);
            registerItem.setVisible(true);
            accessItem.setVisible(false);
        }else{
            loginItem.setTitle(getString(R.string.logout));
            activeItem.setVisible(true);
            registerItem.setVisible(false);
            accessItem.setVisible(true);
        }
    }

    private void applyFontToMenuItem(MenuItem mi) {
        Typeface font = Typeface.createFromAsset(getAssets(), "fonts/Vazir.ttf");
        SpannableString mNewTitle = new SpannableString(mi.getTitle());
        mNewTitle.setSpan(new CustomTypeFaceSpan("", font, Color.WHITE), 0, mNewTitle.length(), Spannable.SPAN_INCLUSIVE_INCLUSIVE);
        mi.setTitle(mNewTitle);
    }



}

File: app/src/main/java/com/kavirelectronic/ali/kavir_info/activity/CategoryActivity.kt
package com.kavirelectronic.ali.kavir_info.activity

import android.app.Fragment
import android.app.FragmentTransaction
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.Typeface
import android.os.Bundle
import com.google.android.material.navigation.NavigationView
import androidx.core.view.GravityCompat
import androidx.drawerlayout.widget.DrawerLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.text.Spannable
import android.text.SpannableString
import android.util.Log
import android.view.Gravity
import android.view.MenuItem
import android.view.View
import android.widget.ImageButton
import android.widget.TextView
import com.kavirelectronic.ali.kavir_info.R
import com.kavirelectronic.ali.kavir_info.activity.CategoryActivity
import com.kavirelectronic.ali.kavir_info.fragments.*
import com.kavirelectronic.ali.kavir_info.utility.CustomTypeFaceSpan
import com.kavirelectronic.ali.kavir_info.utility.FormatHelper
import com.kavirelectronic.ali.kavir_info.utility.SaveItem
import com.kavirelectronic.ali.kavir_info.utility.Utility

class CategoryActivity : AppCompatActivity(), NavigationView.OnNavigationItemSelectedListener {
    private var mTopToolbar: Toolbar? = null
    private var transaction: FragmentTransaction? = null
    private var drawer: DrawerLayout? = null
    private var toolbarTitle: TextView? = null
    private var displayNameTv: TextView? = null
    private var navigationView: NavigationView? = null
    private var searchBtn: ImageButton? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_category)
        context = this
        mTopToolbar = findViewById<View>(R.id.my_toolbar) as Toolbar
        toolbarTitle = findViewById<View>(R.id.category_title) as TextView
        searchBtn = findViewById<View>(R.id.search_btn) as ImageButton
        searchBtn!!.setOnClickListener { startActivity(Intent(this@CategoryActivity, SearchActivity::class.java)) }
        setSupportActionBar(mTopToolbar)
        //// check new version
        if (SaveItem.getItem(this, SaveItem.COME_NEW_VERSION, "").equals("1", ignoreCase = true)) {
            Utility.showAlertNewVersion(this
                    , SaveItem.getItem(this, SaveItem.NEW_VERSION_NAME, "")
                    , SaveItem.getItem(this, SaveItem.NEW_VERSION_URL, ""))
            SaveItem.setItem(this, SaveItem.COME_NEW_VERSION, "0")
        }

        //***************************
        drawer = findViewById<View>(R.id.drawer_layout) as DrawerLayout
        val toggle = ActionBarDrawerToggle(
                this, drawer, mTopToolbar, R.string.navigation_drawer_open, R.string.navigation_drawer_close)
        navigationView = findViewById<View>(R.id.nav_view) as NavigationView
        val headerView = navigationView!!.getHeaderView(0)
        displayNameTv = headerView.findViewById<View>(R.id.display_name_header) as TextView
        var displayName = SaveItem.getItem(this, SaveItem.USER_NAME, "")
        if (displayName == "") {
            displayName = getString(R.string.guest)
        }
        displayNameTv!!.text = FormatHelper.toPersianNumber(displayName)
        navigationView!!.setNavigationItemSelectedListener(this)
        setNameLoginItem()
        toggle.isDrawerIndicatorEnabled = false
        drawer!!.addDrawerListener(toggle)
        toggle.syncState()
        //////***************************************
        findViewById<View>(R.id.drawer_button).setOnClickListener { // open right drawer
            if (drawer!!.isDrawerOpen(GravityCompat.END)) {
                drawer!!.closeDrawer(GravityCompat.END)
            } else drawer!!.openDrawer(GravityCompat.END)
        }
        setFontNavig()
    }

    override fun onResume() {
        try {
            val fragment = fragmentManager.findFragmentById(R.id.fragment) as CategoryFragment
            fragment.dataFromServer()
        } catch (e: Exception) {
            Log.e("onResume:", e.message)
        }
        super.onResume()
    }

    private fun setFontNavig() {
        val m = navigationView!!.menu
        for (i in 0 until m.size()) {
            val mi = m.getItem(i)

            //for aapplying a font to subMenu ...
            val subMenu = mi.subMenu
            if (subMenu != null && subMenu.size() > 0) {
                for (j in 0 until subMenu.size()) {
                    val subMenuItem = subMenu.getItem(j)
                    applyFontToMenuItem(subMenuItem)
                }
            }
            applyFontToMenuItem(mi)
        }
    }

    private fun loadFragment(fragmentNew: Fragment?) {
        transaction = fragmentManager!!.beginTransaction()
        val fragment = fragmentManager!!.findFragmentById(R.id.fragment)
        if (fragment != null) {
            fragmentManager!!.beginTransaction().remove(fragment).commit()
        }
        // load fragment
        transaction!!.replace(R.id.fragment, fragmentNew)
        transaction!!.addToBackStack(null)
        transaction!!.commit()
    }

    override fun onBackPressed() {
        if (drawer!!.isDrawerOpen(GravityCompat.END)) {
            drawer!!.closeDrawer(GravityCompat.END)
        } else {
            super.onBackPressed()
        }
    }

    override fun onNavigationItemSelected(menuItem: MenuItem): Boolean {
        // Handle navigation view item clicks here.
        val id = menuItem.itemId
        if (id == R.id.login) {
            loginProcess()
        } else if (id == R.id.home_page) {
            loadFragment(CategoryFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.home)
            statusSearchBtn(View.VISIBLE)
        } else if (id == R.id.ready_read_page) {
            loadFragment(ReadyReadFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.redy_read)
            statusSearchBtn(View.VISIBLE)
        } else if (id == R.id.about_we) {
            loadFragment(AboutFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.about_we)
            statusSearchBtn(View.VISIBLE)
        } else if (id == R.id.register) {
            loadFragment(RegisterFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.register)
            statusSearchBtn(View.VISIBLE)
        } else if (id == R.id.active) {
            loadFragment(ActiveFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.active)
            statusSearchBtn(View.VISIBLE)
        } else if (id == R.id.access) {
            loadFragment(AccessFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.access_list)
            statusSearchBtn(View.GONE)
        } else if (id == R.id.help) {
            loadFragment(HelpFragment.newInstance())
            toolbarTitle!!.text = getString(R.string.help)
            statusSearchBtn(View.VISIBLE)
        }
        drawer!!.closeDrawer(GravityCompat.END)
        return true
    }

    private fun statusSearchBtn(visible: Int) {
        searchBtn!!.visibility = visible
    }

    private fun loginProcess() {
        val menu = navigationView!!.menu
        val loginItem = menu.findItem(R.id.login)
        if (loginItem.title.toString().equals(getString(R.string.login), ignoreCase = true)) {
            startActivity(Intent(this@CategoryActivity, LoginActivity::class.java))
        } else {
            logout()
        }
    }

    private fun logout() {
        clearSeesion()
        val intent = intent
        finishAffinity()
        startActivity(intent)
    }

    private fun clearSeesion() {
        SaveItem.setItem(this, SaveItem.USER_FIRST_NAME, "")
        SaveItem.setItem(this, SaveItem.USER_LAST_NAME, "")
        SaveItem.setItem(this, SaveItem.USER_EMAIL, "")
        SaveItem.setItem(this, SaveItem.USER_NAME, "")
        SaveItem.setItem(this, SaveItem.USER_MOBILE, "")
        SaveItem.setItem(this, SaveItem.USER_ID, "")
        SaveItem.setItem(this, SaveItem.USER_COOKIE, "")
    }


    private fun setNameLoginItem() {
        val menu = navigationView!!.menu
        val loginItem = menu.findItem(R.id.login)
        val activeItem = menu.findItem(R.id.active)
        val registerItem = menu.findItem(R.id.register)
        val accessItem = menu.findItem(R.id.access)
        if (SaveItem.getItem(this, SaveItem.USER_COOKIE, "") == "") {
            loginItem.title = getString(R.string.login)
            activeItem.isVisible = false
            registerItem.isVisible = true
            accessItem.isVisible = false
        } else {
            loginItem.title = getString(R.string.logout)
            activeItem.isVisible = true
            registerItem.isVisible = false
            accessItem.isVisible = true
        }
    }

    private fun applyFontToMenuItem(mi: MenuItem) {
        val font = Typeface.createFromAsset(assets, "fonts/Vazir.ttf")
        val mNewTitle = SpannableString(mi.title)
        mNewTitle.setSpan(CustomTypeFaceSpan("", font, Color.WHITE), 0, mNewTitle.length, Spannable.SPAN_INCLUSIVE_INCLUSIVE)
        mi.title = mNewTitle
    }

    companion object {
        @JvmField
        var context: Context? = null
    }
}